backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt

```diff
@@ -96,22 +94,36 @@ internal fun IrFunction.hasReferenceAt(index: Int): Boolean {
     }
 }
 
-private fun IrFunction.needBridgeToAt(target: IrFunction, index: Int)
-        = hasValueTypeAt(index) xor target.hasValueTypeAt(index)
+private fun IrFunction.needBridgeToAt(target: IrFunction, index: Int) =
+        bridgeDirectionToAt(target, index) != BridgeDirection.NOT_NEEDED
 
 internal fun IrFunction.needBridgeTo(target: IrFunction)
         = (0..this.valueParameters.size + 2).any { needBridgeToAt(target, it) }
 
-internal enum class BridgeDirection {
-    NOT_NEEDED,
-    FROM_VALUE_TYPE,
-    TO_VALUE_TYPE
+internal sealed class BridgeDirection {
+    object NOT_NEEDED : BridgeDirection()
+    object FROM_VALUE_TYPE : BridgeDirection()
+    object TO_VALUE_TYPE : BridgeDirection()
+    object TO_NULL : BridgeDirection()
+    data class TO_NOTHING(val fromBinaryType: PrimitiveBinaryType?) : BridgeDirection()
```

Comment:
Do we really need all these kinds?
We already have generic machinery that supports arbitrary upcasting and downcasting.
So why is it required to distinguish all these cases in bridge generator, isn't it enough to have only "needed" and "not needed"?

Reply:
The original idea was that a bridge could be described by a pair `<function, bridgeDirections>`, and that one bridge could be used to override more than one function. Note that `bridgeDirections` is used as a key in the vtable generator, so I tried not to touch that. We can discuss the refactoring of the whole bridge/vtable system but I think it is a separate thing and IMO not for 1.4-M3.

Reply:
> and IMO not for 1.4-M3.

This entire PR is not for 1.4-M3 or 1.4 anyway.
I find this classification of bridge directions too cumbersome, I believe it is possible to make it much simpler without refactoring the whole system.

Reply:
Ok, if it is not intended to be fixed in 1.4, then, sure, I'll try to simplify things a bit.